### PR TITLE
Use standalone language server

### LINKED ISSUE #264: The matched }/]/) doesn't appear when trying to type {/[/( in the .json file
**Extension Build:** [0.6.1-alphalangserv](\\scratch2\scratch\stephwe\ARM\VSIX\2019-08-15a-json-completion)
**OS**: All

**Steps to reproduce:**
1. Open one json file then type '{'
2. Check whether the matched '}' appears or not.

**Expect Experience:**

The matched '}' appears.

**Actual Experience:**

The matched '}' doesn't appear.

**More info:**

This issue also reproduces when typing [/(/'/".

### LINKED ISSUE #269: The matching end "}" isn't selected when selecting the begining "{"
**OS**: Mac & Linux & Windows
**Build Version**: [0.6.1-alphalangserv](\\scratch2\scratch\stephwe\ARM\VSIX\2019-08-15a-json-completion)
**Regression**: Not a Regression

**Repro Steps:**
1. Open one deployment file in VS Code -->Select one beginning "{" in editor.
2. Check whether the matching end "}" is selected or not.

**Expect:**
The matching end "}" is selected.

**Actual:**
he matching end "}" isn't selected.

**More Info:**
1. The matching "]"/"[" is selected when selecting another one.
2. The matching ")"/"(" is selected when selecting another one.
3. The beginning "{" is also not selected when selecting the ending "}".

### LINKED ISSUE #270: "Go to Bracket" item doesn't work
**OS**: Mac & Linux & Windows
**Build Version**: [0.6.1-alphalangserv](\\scratch2\scratch\stephwe\ARM\VSIX\2019-08-15a-json-completion)
**Regression**: Not a Regression

**Repro Steps:**
1. Open one deployment file --> Select the beginning "{" -->Go --> Go to Bracket.
2. Check the result.

**Expect:**
Edit should select the end "}".

**Actual:**
The "Go to Bracket" item doesn't work.

### LINKED ISSUE #273: The selected characters are removed when typing an apostrophe
**OS**: Mac & Linux & Windows
**Build Version**: [0.6.1-alphalangserv](\\scratch2\scratch\stephwe\ARM\VSIX\2019-08-15a-json-completion)
**Regression**: Not a Regression

**Repro Steps:**
1. Open one deployment file -->Double click to select a set of characters insider a string express.
2. Type an apostrophe (').
3. Check the result.

**Expect:**
The set of characters selected is surrounded by on both sides by an apostrophe.
image

**Actual:**
The selected characters are removed not surrounded by on both sides by an apostrophe.

### PR DESCRIPTION
**Pull request by StephenWeatherford**
NOTE: Tests will not currently pass because the language server isn’t being picked up in the build yet.  Will need to figure out a solution before merging, or maybe I'll temporarily disable some functional tests.

The large majority of file changes are test cases and test code.

1. Add new language ID ‘arm-template’
1. Switch language if ARM schema detected
1. To simplify, do not unhook events once we’ve hooked them up (didn’t provide much value)
1. Remove the injected grammar which detects the ARM schema - grammar is now completely controlled by language id
1. Add some telemetry about usage of params/vars/namespaces usage
1. Integrate new ARM language server (to replace our current usage of the JSON language server)
1. Rather than parsing every JSON document on every edit to determine if it’s an ARM template, instead use regex on the first part of the file (like VS does)
1. Unit and functional tests

Fixes #264 
Fixes #269 
Fixes #270 
Fixes #273

### CODE REVIEW

**Comment by StephenWeatherford** [comment]
I guess I won't be able to merge until we get a published version of the language server bits that I can pull in to the extension during build.  But I'd still like to get this branch reviewed in the meantime so it's not all piled up until the end.  I've also got a few more changes coming, so I'll do PRs into this branch for now.

File: src/languageclient/startArmLanguageServer.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        }
+        if (ext.addCompletionDiagnostic) {
+            // Forces the server to add a completion message to its diagnostics
+            commonArgs.push('--verbose-diagnostics');
+        }
+
+        // If the extension is launched in debug mode then the debug server options are used
+        // Otherwise the run options are used
+        let serverOptions: ServerOptions = {
+            run: {
```
Comment (nturinski) [inline src/languageclient/startArmLanguageServer.ts]: nit: Inconsistent format

File: src/Treeview.ts [hunk 2]
```diff
@@ -70,7 +69,7 @@ export class JsonOutlineProvider implements vscode.TreeDataProvider<string> {
     public getChildren(element?: string): string[] {
         // check if there is a visible text editor
         if (vscode.window.visibleTextEditors.length > 0) {
-            if (isLanguageIdSupported(vscode.window.activeTextEditor.document.languageId)) {
+            if (vscode.window.activeTextEditor && this.isArmTemplateDocument(vscode.window.activeTextEditor.document)) {
```
Comment (nturinski) [inline src/Treeview.ts]: Is checking for the `activeTextEditor` necessary here?  I'm just asking since this is a newly introduced behavior.

Comment (StephenWeatherford) [reply src/Treeview.ts]: I dereference it in the next part of the expression, so yes.

File: src/languageclient/startArmLanguageServer.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+export function startArmLanguageServer(context: ExtensionContext): void {
+    callWithTelemetryAndErrorHandlingSync('startArmLanguageClient', () => {
+        // The server is implemented in .NET Core. We run it by calling 'dotnet' with the dll as an argument
+
+        let serverExe = os.platform() === 'win32' ? 'dotnet.exe' : 'dotnet';
+
+        let serverDllPath = workspace.getConfiguration('armTools').get<string | undefined>('languageServer.path');
+
+        if (typeof serverDllPath !== 'string' || serverDllPath === '') {
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/languageclient/startArmLanguageServer.ts]: You would think, wouldn't you?  But vscode gives you back whatever's in the settings, even if it's an object or number or whatever.

**Comment by StephenWeatherford** [comment]
Thanks Nathan!

**Comment by StephenWeatherford** [comment]
Failing tests are temporarily disabled.

**Comment by StephenWeatherford** [comment]
NOTE: This PR is still active, but waiting on resolution of some licensing/build issues with the language server bits before I can check it in.

**Pull request merged**